File: src/main/scala/jk_5/nailed/server/team/NailedTeam.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.team

import jk_5.nailed.api.map.Team
import jk_5.nailed.api.mappack.metadata.MappackTeam
import jk_5.nailed.api.player.Player
import jk_5.nailed.api.util.Location
import jk_5.nailed.server.map.TeamManager

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
class NailedTeam(val mappackTeam: MappackTeam, val manager: TeamManager) extends Team {

  private val memberSet = mutable.HashSet[Player]()
  var spawnpoint: Location = null

  override val id = mappackTeam.id
  override val color = mappackTeam.color
  override val name = mappackTeam.name
  override def getName = name

  val scoreboardTeam = manager.getScoreboardManager.getOrCreateTeam(this.id)
  scoreboardTeam.setPrefix(color.toString)
  scoreboardTeam.setDisplayName(name)

  override def members = java.util.Arrays.asList(this.memberSet.toArray: _*)

  def onPlayerJoined(player: Player){
    memberSet += player
    scoreboardTeam.addPlayer(player)
  }

  def onPlayerLeft(player: Player){
    memberSet -= player
    scoreboardTeam.removePlayer(player)
  }

  override def getSpawnPoint = spawnpoint
  override def setSpawnPoint(spawnpoint: Location) = this.spawnpoint = spawnpoint
}

File: src/main/java/jk_5/nailed/server/team/NailedTeam.java
package jk_5.nailed.server.team;

import com.google.common.collect.ImmutableSet;
import jk_5.nailed.api.chat.ChatColor;
import jk_5.nailed.api.map.Team;
import jk_5.nailed.api.mappack.metadata.MappackTeam;
import jk_5.nailed.api.player.Player;
import jk_5.nailed.api.scoreboard.ScoreboardTeam;
import jk_5.nailed.api.util.Location;
import jk_5.nailed.server.map.MapTeamManager;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

public class NailedTeam implements Team {

    private final MappackTeam mappackTeam;
    private final Set<Player> memberSet = new HashSet<Player>();
    private final ScoreboardTeam scoreboardTeam;
    private Location spawnpoint = null;

    public NailedTeam(MappackTeam mappackTeam, MapTeamManager manager) {
        this.mappackTeam = mappackTeam;
        this.scoreboardTeam = manager.getMap().getScoreboardManager().getOrCreateTeam(mappackTeam.id());
        this.scoreboardTeam.setPrefix(mappackTeam.color().toString());
        this.scoreboardTeam.setDisplayName(mappackTeam.name());
    }

    public void onPlayerJoined(@Nonnull Player player){
        this.memberSet.add(player);
        scoreboardTeam.addPlayer(player);
    }

    public void onPlayerLeft(@Nonnull Player player){
        this.memberSet.remove(player);
        this.scoreboardTeam.removePlayer(player);
    }

    @Nonnull
    @Override
    public String id() {
        return mappackTeam.id();
    }

    @Nonnull
    @Override
    public String name() {
        return mappackTeam.name();
    }

    @Nonnull
    @Override
    public ChatColor color() {
        return mappackTeam.color();
    }

    @Nonnull
    @Override
    public Collection<Player> members() {
        return ImmutableSet.copyOf(memberSet);
    }

    @Nullable
    @Override
    public Location getSpawnPoint() {
        return spawnpoint;
    }

    @Override
    public void setSpawnPoint(@Nullable Location spawnpoint) {
        this.spawnpoint = spawnpoint;
    }

    @Override
    public String getName() {
        return mappackTeam.name();
    }
}
